primitives/constraints/ssz.go

```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: BUSL-1.1
+//
+// Copyright (C) 2025, Berachain Foundation. All rights reserved.
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file of this repository and at www.mariadb.com/bsl11.
+//
+// ANY USE OF THE LICENSED WORK IN VIOLATION OF THIS LICENSE WILL AUTOMATICALLY
+// TERMINATE YOUR RIGHTS UNDER THIS LICENSE FOR THE CURRENT AND ALL OTHER
+// VERSIONS OF THE LICENSED WORK.
+//
+// THIS LICENSE DOES NOT GRANT YOU ANY RIGHT IN ANY TRADEMARK OR LOGO OF
+// LICENSOR OR ITS AFFILIATES (PROVIDED THAT YOU MAY USE A TRADEMARK OR LOGO OF
+// LICENSOR AS EXPRESSLY REQUIRED BY THIS LICENSE).
+//
+// TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+// AN “AS IS” BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+// EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+// TITLE.
+
+package constraints
+
+import (
+	"github.com/berachain/beacon-kit/primitives/common"
+)
+
+// SSZMarshaler is an interface for objects that can be
+// marshaled to SSZ format.
+type SSZMarshaler interface {
+	// MarshalSSZ marshals the object into SSZ format.
+	MarshalSSZ() ([]byte, error)
+}
+
+// SSZUnmarshaler is an interface for objects that can be unmarshaled from SSZ format.
+type SSZUnmarshaler[SelfT any] interface {
+	// NewFromSSZ creates a new object from SSZ format.
+	NewFromSSZ(bz []byte) (SelfT, error)
```

Comment:
just a note, feel free to neglect this.
Looks like practically we always call `NewFromSSZ`. It would be great to make a templated constructor out of this, dropping the pointer receiver.
I can't recall rn if golang supports generics specialization so not sure if this is feasible.
My whole point is that the pointer receiver seems to exist only to allow us picking the right constructor

Reply:
as discussed this is necessary for the generic (en/decoders) [here](https://github.com/berachain/beacon-kit/blob/d6ad0abe380b1f89d58e6d3c29437e1fcac48180/storage/encoding/ssz.go#L42) and [here](https://github.com/berachain/beacon-kit/blob/d6ad0abe380b1f89d58e6d3c29437e1fcac48180/storage/encoding/ssz.go#L93). To completely remove the generics we could build a en/decoder for the cosmos store for each type that needs it (only like 6 or 7 of them). Open to it, but should be subsequent PR

Reply:
Indeed done with two PRs https://github.com/berachain/beacon-kit/pull/2615 and https://github.com/berachain/beacon-kit/pull/2616